### ISSUE TITLE
Derivation details of reg: gamma

### ISSUE THREAD

**Issue by HowardHChuang** [dmlc/xgboost#2826]
Dear all,

I recently adopt xgboost with reg:gamma to tackle a demand forecasting in high-tech electronics sectors (with preliminary success), and intend to extend the "reg: gamma" into more generic cases, e.g., generalize gamma quantile regression. 

Are there formal references on how the loss functions, gradient, & hessian for the "reg: gamma" are derived? Any comments/suggestions will be appreciated.

Howard
